src/core/cgroup.c

```diff
@@ -3510,6 +3510,35 @@ void manager_invalidate_startup_units(Manager *m) {
                 unit_invalidate_cgroup(u, CGROUP_MASK_CPU|CGROUP_MASK_IO|CGROUP_MASK_BLKIO);
 }
 
+int compare_job_cpu_weight(const void *a, const void *b)
+{
+        Job *x = (Job*)a, *y = (Job*)b;
+        ManagerState state = manager_state(x->unit->manager);
+        CGroupContext *context;
+        uint64_t weight_x, weight_y;
+        int ret;
+
+        context = unit_get_cgroup_context(x->unit);
+        weight_x = context ? cgroup_context_cpu_weight(context, state) : CGROUP_WEIGHT_DEFAULT;
+        context = unit_get_cgroup_context(y->unit);
+        weight_y = context ? cgroup_context_cpu_weight(context, state) : CGROUP_WEIGHT_DEFAULT;
+
+        /* Sort jobs for all units by CPU weigh:
+         * - units that fork new processes after any other units
+         * - units with higher CPU weight first
+         * - sort by id if the other criteria are the same */
+        if (x->unit->type > UNIT_SWAP && y->unit->type <= UNIT_SWAP)
+                return -1;
+
+        if (x->unit->type <= UNIT_SWAP && y->unit->type > UNIT_SWAP)
+                return 1;
+
+        if (x->unit->type <= UNIT_SWAP && (ret = CMP(weight_y, weight_x)) != 0)
```

Comment:
I think this check `x->unit->type <= UNIT_SWAP` is unnecessary: if we set the context and weight, we should just use it.

Reply:
Yes of course. Both weights are `CGROUP_WEIGHT_DEFAULT` in that case, so the `strcmp` fallback is still used.